graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/internal/DgsSchemaProvider.kt

```diff
@@ -154,15 +155,16 @@ class DgsSchemaProvider(
     }
 
     private fun invokeDgsTypeDefinitionRegistry(dgsComponent: Any, registry: TypeDefinitionRegistry): TypeDefinitionRegistry? {
-        return dgsComponent.javaClass.methods.filter { it.isAnnotationPresent(DgsTypeDefinitionRegistry::class.java) }
+        return dgsComponent.javaClass.methods.asSequence()
```

Comment:
[filter](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/filter.html) and [map](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/map.html) both return lists, but at the end of the chain it's a `reduceOrNull` call anyway so those intermediate lists are just wasteful in terms of allocations.

To put it another way, if it were unrolled into an imperative style, it would be the equivalent of doing this:

```kotlin
val filtered = mutableListOf<Method>()
for (method in dgsComponent.javaClass.methods) {
    if (method.isAnnotationPresent(DgsTypeDefinitionRegistry::class.java)) {
        filtered += method
    }
}

val registries = mutableListOf<TypeDefinitionRegistry>()
for (method in filtered) {
    registries += method.invoke(/* ... */) as TypeDefinitionRegisty
}

var typeDefRegistry: TypeDefinitionRegistry? = null
for (registry in registries) {
    typeDefRegistry = typeDefRegistry?.merge(registry) ?: registry
}
return registry
```

In fact if you look at the generated bytecode, it's likely doing something similar under the hood. Putting `asSequence()` in there as the first step just means it's all done in 1 shot, so things are not iterated 3 times, and the intermediate List allocations don't even need to happen.

Reply:
TIL, thanks for the explanation.